core/client/src/main/java/alluxio/client/block/UnderFileSystemBlockInStream.java

```diff
@@ -0,0 +1,226 @@
+import alluxio.wire.WorkerNetAddress;
+import alluxio.worker.block.io.LocalFileBlockReader;
+
+import com.codahale.metrics.Counter;
+import com.google.common.io.Closer;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class provides a streaming API to read a UFS block from an Alluxio worker. The data will
+ * be transferred through an Alluxio worker's dataserver to the client.
+ */
+@NotThreadSafe
+public final class UnderFileSystemBlockInStream extends BufferedBlockInStream implements Locatable {
+  /** Used to manage closeable resources. */
+  private final Closer mCloser;
+  /** If set, the block will not be cached into Alluxio because of this read. */
+  private final boolean mNoCache;
+  /** Whether we are reading from a local Alluxio worker or a remote one. */
+  private final boolean mLocal;
+
+  /** Client to communicate with the remote worker. */
+  private final BlockWorkerClient mBlockWorkerClient;
+  /** The file system context which provides block worker clients. */
+  private final FileSystemContext mContext;
+  /** {@link UnderFileSystemBlockReader} for this instance. */
+  private UnderFileSystemBlockReader mReader;
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockInStream}. It keeps polling the worker
+   * until the block is cached to Alluxio or it successfully acquires a UFS read token. If
+   * the block is cached to Alluxio, it returns an {@link BufferedBlockInStream} that reads from
+   * Alluxio.
+   *
+   * @param context the file system context
+   * @param ufsPath the UFS path
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockStart the start position of the block in the UFS file
+   * @param workerNetAddress the worker network address
+   * @param options the in stream options
+   * @return the {@link UnderFileSystemBlockInStream} instance or the {@link BufferedBlockOutStream}
+   *         that reads from Alluxio directly
+   * @throws IOException if it fails to create {@link UnderFileSystemBlockInStream}
+   */
+  public static BufferedBlockInStream create(FileSystemContext context, String ufsPath,
+      long blockId, long blockSize, long blockStart, WorkerNetAddress workerNetAddress,
+      InStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createBlockWorkerClient(workerNetAddress));
+      LockBlockOptions lockBlockOptions =
+          LockBlockOptions.defaults().setUfsPath(ufsPath).setOffset(blockStart)
+              .setBlockSize(blockSize).setMaxUfsReadConcurrency(options.getMaxUfsReadConcurrency());
+      LockBlockResult result = blockWorkerClient.lockUfsBlock(blockId, lockBlockOptions);
+      closer.register(new LockBlockResource(blockWorkerClient, blockId));
```

Comment:
This seems fairly awkward since we are creating an object for the sole purpose of submitting it to the closer.